### DISCUSSION TITLE
So you want to build py5 from source to test the pre-release (or hack it yourself)

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#124] [Show and tell]
Start here: https://py5.ixora.io/developer/build_process.html

I had some issues with the Java version, and @hx2A is helping me sort things out!

**Comment by villares** [comment]
Right now:
```
$ make processing_dir=~/other-software/processing-4.0.1/ py5_build_dir=../py5code skip_autopep8=true
ant -f py5_jar/build.xml -Dprocessing_dir=/home/villares/other-software/processing-4.0.1
Buildfile: /home/villares/GitHub/pythonprocessing/py5generator/py5_jar/build.xml

[http://project.run](https://t.co/7E69LNtKLj):

compile:
    [javac] Compiling 10 source files to .../pythonprocessing/py5generator/py5_jar/bin
    [javac] error: invalid target release: 17
    [javac] Usage: javac <options> <source files>
    [javac] use --help for a list of possible options

BUILD FAILED
.../pythonprocessing/py5generator/py5_jar/build.xml:13: The following error occurred while executing this line:
.../pythonprocessing/py5generator/py5_jar/build.xml:19: Compile failed; see the compiler error output for details.
```
My `java -version`  told me: `openjdk version "18.0.2"`
But `javac - version` looks like trouble:  `javac 11.0.15-internal`

**Reply by hx2A** [reply]
Can you try all of these commands?

```bash
(py5) [jim@main py5generator]$ ant -version
Apache Ant(TM) version 1.10.12 compiled on June 6 2022
(py5) [jim@main py5generator]$ java -version
openjdk version "17.0.4" 2022-07-19
OpenJDK Runtime Environment (Red_Hat-17.0.4.0.8-1.fc36) (build 17.0.4+8)
OpenJDK 64-Bit Server VM (Red_Hat-17.0.4.0.8-1.fc36) (build 17.0.4+8, mixed mode, sharing)
(py5) [jim@main py5generator]$ javac -version
javac 17.0.4
(py5) [jim@main py5generator]$ 
```

I am going to guess your Java installation is a Java Runtime Environment (JRE) and not a (SDK) which contains the ability to compile Java and not just run it.

**Reply by villares** [reply]
```
(py5) ... $ ant -version
Apache Ant(TM) version 1.10.12 compiled on February 5 2022

(py5) ... $ java -version
openjdk version "18.0.2" 2022-07-19
OpenJDK Runtime Environment (build 18.0.2+9)
OpenJDK 64-Bit Server VM (build 18.0.2+9, mixed mode)

(py5) ... $ javac -version
javac 11.0.15-internal
```

**Reply by hx2A** [reply]
How did you install the Java stuff on this machine? Was it the [install-jdk](https://github.com/jyksnw/install-jdk) library?

Review the usage of install-jdk. Try this:

```bash
jdk.install('17', jre=False)
```
It should install to a `~/.jdk` directory, not a `~/.jre` directory.

It seems that when Thonny installs Java, [it installs the JDK](https://github.com/tabreturn/thonny-py5mode/blob/a9843ba058d72be0f49b5ff6b8abe21c90cb6e76/thonnycontrib/thonny-py5mode/install_jdk.py#L53), which is what you want in this case.

As a side note, is installing the JDK the right choice for most of our users? Wouldn't Thonny work just fine with the JRE instead? I don't want to take up too much of people's hard drives for no reason. (I believe the JDK is larger than the JRE)

**Reply by villares** [reply]
oh dear, I have no idea how I've got here... I have more than one JDK-17 on this machine, I'll try to point the conda env to the other one... outside the env I get `javac 18.0.2` :-S
Trying to extrincate openjdk from conda...
I'm afraid I'll want to reset this whole conda thing because it uninstalled `ant` and is complaining about incosistent env...

**Reply by hx2A** [reply]
OK, you need to uninstall Java from the anaconda environment. I'm afraid Anaconda is a bad way to install Java now that py5 requires Java 17 or higher.

Or if this is a new environment, try deleting it and make a new one from scratch.

**Reply by villares** [reply]
Some progress! https://gist.github.com/villares/d415e11f9c9ca291f98d6635a97b8817

```
]$ make processing_dir=~/other-software/processing-4.0.1/ py5_build_dir=../py5code skip_autopep8=true
ant -f py5_jar/build.xml -Dprocessing_dir=/home/villares/other-software/processing-4.0.1
Buildfile: /home/villares/GitHub/pythonprocessing/py5generator/py5_jar/build.xml

project.run:

compile:
    [javac] Compiling 10 source files to /home/villares/GitHub/pythonprocessing/py5generator/py5_jar/bin
    [javac] warning: [options] system modules path not set in conjunction with -source 17
    [javac] 1 warning

dist:
    [mkdir] Created dir: /home/villares/GitHub/pythonprocessing/py5generator/py5_jar/dist
      [jar] Building jar: /home/villares/GitHub/pythonprocessing/py5generator/py5_jar/dist/py5.jar

cleanup:
   [delete] Deleting directory /home/villares/GitHub/pythonprocessing/py5generator/py5_jar/bin

BUILD SUCCESSFUL
Total time: 1 second
python generate_py5.py /home/villares/other-software/processing-4.0.1/ ../py5code --skip_autopep8
2022-08-15 16:27:48,197 | WARNING | skipping autopep8 formatting
2022-08-15 16:27:48,197 | INFO | generating py5 library...
...

...
Building wheels for collected packages: py5
  Building wheel for py5 (setup.py) ... done
  Created wheel for py5: filename=py5-0.8.2.dev0-py3-none-any.whl size=11455896 sha256=b8ee7158e45104e41b8357eda04ee6f4322f5ff7d29d76d31534d8b7200ce1b0
  Stored in directory: /tmp/pip-ephem-wheel-cache-5vmc2l8u/wheels/3b/3a/fd/27e4308c750bc946d6234e5aadc4603d55f6394a0d683b5b11
Successfully built py5
Installing collected packages: py5
Successfully installed py5-0.8.2.dev0
touch ../py5code/.install_py5.nogit
```
I have removed the jdk-11 from the anaconda env, installed `ant` from Manjaro interface, and it started seeing `javac 18` and now it worked?

**Reply by hx2A** [reply]
> `BUILD SUCCESSFUL`

It worked! Good job!

**Reply by villares** [reply]
![image](https://user-images.githubusercontent.com/3694604/184705625-ed180327-c1f0-4425-a814-e726eea30773.png)
(I had reverted to 0.8.0a to test if that setup space bug would disappear)

**Reply by hx2A** [reply]
That setup space bug was pretty strange. I'm still convinced it is was problem with the Python standard library but I couldn't pin it down. In any case, it should work correctly now.

Thonny is using a miniconda environment? I didn't know Thonny could work that way. I thought it was using a virtual environment it managed itself. When I was testing I went through Thonny's package manager and told it to install a Python wheel that was available locally.

**Reply by villares** [reply]
yeah, you can point to the interpreter here:
![image](https://user-images.githubusercontent.com/3694604/184787045-a7b76639-e545-4feb-a874-98f12e3c10b3.png)
Sometimes strange things happen if Thonny itself is running on a different interpreter, but mostly it works.
